In [48]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
print("Libraries imported successfully")

Libraries imported successfully


In [49]:
df = pd.read_csv(r"used_car_dataset.csv")
print(" Dataset loaded successfully")

 Dataset loaded successfully


In [51]:
sns.set_style('darkgrid')
plt.rcParams['figure.figsize'] = (12, 6)

In [54]:
print("First 5 rows:")
df.head()

First 5 rows:


,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"


In [56]:
print("Data Types:")
df.info()

Data Types:
<class 'pandas.DataFrame'>
RangeIndex: 9582 entries, 0 to 9581
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Brand         9582 non-null   str  
 1   model         9582 non-null   str  
 2   Year          9582 non-null   int64
 3   Age           9582 non-null   int64
 4   kmDriven      9535 non-null   str  
 5   Transmission  9582 non-null   str  
 6   Owner         9582 non-null   str  
 7   FuelType      9582 non-null   str  
 8   PostedDate    9582 non-null   str  
 9   AdditionInfo  9582 non-null   str  
 10  AskPrice      9582 non-null   str  
dtypes: int64(2), str(9)
memory usage: 823.6 KB


In [57]:
df['AskPrice'] = df['AskPrice'].astype(str).str.replace(r'[₹,\s]', '', regex=True)
df['AskPrice'] = pd.to_numeric(df['AskPrice'], errors='coerce')

In [63]:
df['kmDriven'] = df['kmDriven'].astype(str).str.replace(r'[,]', '', regex=True)  # Remove commas first
df['kmDriven'] = df['kmDriven'].str.replace(r'[kKmM]', '', regex=True)  # Remove k or m (case insensitive)
df['kmDriven'] = df['kmDriven'].str.strip()  # Remove any extra spaces
df['kmDriven'] = pd.to_numeric(df['kmDriven'], errors='coerce')

In [65]:
print(f"✅ AskPrice converted to {df['AskPrice'].dtype}")
print(f"   Min: {df['AskPrice'].min():,.0f}, Max: {df['AskPrice'].max():,.0f}")
print(f"   Rows with invalid prices: {df['AskPrice'].isnull().sum()}\n")

# Convert kmDriven - Remove km, commas, and spaces
print(f"✅ KmDriven converted to {df['kmDriven'].dtype}")
print(f"   Min: {df['kmDriven'].min():,.0f}, Max: {df['kmDriven'].max():,.0f}")
print(f"   Rows with invalid prices: {df['kmDriven'].isnull().sum()}\n")

✅ AskPrice converted to int64
   Min: 15,000, Max: 42,500,000
   Rows with invalid prices: 0

✅ KmDriven converted to float64
   Min: 0, Max: 980,002
   Rows with invalid prices: 47



In [66]:
print(f"\n📊 Dataset Size: {df.shape[0]:,} rows, {df.shape[1]} columns")

print("\n📋 Column Details:")
for col in df.columns:
    dtype = df[col].dtype.name
    missing = df[col].isnull().sum()
    unique = df[col].nunique()
    print(f"  {col:20s} | Type: {dtype:12s} | Missing: {missing:5d} | Unique: {unique}")


📊 Dataset Size: 9,582 rows, 11 columns

📋 Column Details:
  Brand                | Type: str          | Missing:     0 | Unique: 39
  model                | Type: str          | Missing:     0 | Unique: 400
  Year                 | Type: int64        | Missing:     0 | Unique: 32
  Age                  | Type: int64        | Missing:     0 | Unique: 32
  kmDriven             | Type: float64      | Missing:    47 | Unique: 1745
  Transmission         | Type: str          | Missing:     0 | Unique: 2
  Owner                | Type: str          | Missing:     0 | Unique: 2
  FuelType             | Type: str          | Missing:     0 | Unique: 3
  PostedDate           | Type: str          | Missing:     0 | Unique: 12
  AdditionInfo         | Type: str          | Missing:     0 | Unique: 7307
  AskPrice             | Type: int64        | Missing:     0 | Unique: 1330


In [67]:
print("📈 Quick Statistics:")
print(f"  Average Price: ₹{df['AskPrice'].mean():,.0f}")
print(f"  Price Range: ₹{df['AskPrice'].min():,.0f} - ₹{df['AskPrice'].max():,.0f}")
print(f"  Average Age: {df['Age'].mean():.1f} years")
print(f"  Average KM: {df['kmDriven'].mean():,.0f}")
print(f"  Most Common Brand: {df['Brand'].mode()[0]}")
print(f"  Most Common Fuel: {df['FuelType'].mode()[0]}")

📈 Quick Statistics:
  Average Price: ₹1,063,609
  Price Range: ₹15,000 - ₹42,500,000
  Average Age: 7.6 years
  Average KM: 70,606
  Most Common Brand: Maruti Suzuki
  Most Common Fuel: Diesel


In [69]:
numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = df.select_dtypes(include=['object', 'string']).columns.tolist()

print("Numerical Columns:", numerical_cols)
print("\nCategorical Columns:", categorical_cols)

Numerical Columns: ['Year', 'Age', 'kmDriven', 'AskPrice']

Categorical Columns: ['Brand', 'model', 'Transmission', 'Owner', 'FuelType', 'PostedDate', 'AdditionInfo']


 Day 1 Summary
 
 ### What I Did Today:
 - ✅ Loaded the used car dataset
 - ✅ Cleaned AskPrice and kmDriven columns (removed ₹, commas, spaces)
 - ✅ Converted them to numeric types
 - ✅ Explored data types and structure
 - ✅ Identified missing values
 - ✅ Calculated basic statistics
 - ✅ Separated numerical and categorical columns
 
 ### Key Observations:
 1. **Dataset Size**: 9,582 car listings with 11 features
 2. **Missing Data**: kmDriven and AdditionalInfo have missing values
 3. **Price Range**: Very wide (from budget to luxury cars)
 4. **Most Common Brand**: Maruti Suzuki dominates the market
 5. **Age Column**: Already exists (0-38 years)
 
 ### Next Steps (Day 2):
 - Clean missing values
 - Handle outliers in price
 - Remove duplicates
 - Prepare data for analysis